src/include/btree_inline.h

```diff
@@ -1028,53 +1048,86 @@ __wt_row_leaf_key_set_cell(WT_PAGE *page, WT_ROW *rip, WT_CELL *cell)
 static inline void
 __wt_row_leaf_key_set(WT_PAGE *page, WT_ROW *rip, WT_CELL_UNPACK_KV *unpack)
 {
-    uintptr_t v;
+    uintptr_t key_offset, v;
 
     /*
      * See the comment in __wt_row_leaf_key_info for an explanation of the magic.
+     *
+     * Not checking the prefix and cell offset sizes, the fields hold any legitimate value.
      */
-    v = WT_K_ENCODE_KEY_LEN(unpack->size) |
-      WT_K_ENCODE_KEY_OFFSET(WT_PAGE_DISK_OFFSET(page, unpack->data)) | WT_K_FLAG;
-    WT_ASSERT(NULL, WT_ROW_SLOT(page, rip) < page->entries);
+    key_offset = (uintptr_t)WT_PTRDIFF(unpack->cell, unpack->data);
+    if (unpack->type != WT_CELL_KEY || key_offset > WT_K_MAX_KEY_OFFSET ||
+      unpack->size > WT_K_MAX_KEY_LEN)
+        v = WT_CELL_ENCODE_OFFSET(WT_PAGE_DISK_OFFSET(page, unpack->cell)) | WT_CELL_FLAG;
+    else
+        v = WT_K_ENCODE_KEY_CELL_OFFSET(WT_PAGE_DISK_OFFSET(page, unpack->cell)) |
+          WT_K_ENCODE_KEY_PREFIX(unpack->prefix) | WT_K_ENCODE_KEY_OFFSET(key_offset) |
+          WT_K_ENCODE_KEY_LEN(unpack->size) | WT_K_FLAG;
+
     WT_ROW_KEY_SET(rip, v);
 }
 
 /*
  * __wt_row_leaf_value_set --
- *     Set a WT_ROW to reference an on-page row-store leaf value.
+ *     Set a WT_ROW to reference an on-page row-store leaf key and value pair, if possible.
  */
 static inline void
-__wt_row_leaf_value_set(WT_PAGE *page, WT_ROW *rip, WT_CELL_UNPACK_KV *unpack)
+__wt_row_leaf_value_set(WT_ROW *rip, WT_CELL_UNPACK_KV *unpack)
 {
-    uintptr_t key_len, key_offset, value_offset, v;
+    uintptr_t value_offset, v;
 
+    /* The row-store key can change underfoot; explicitly take a copy. */
     v = (uintptr_t)WT_ROW_KEY_COPY(rip);
 
     /*
      * See the comment in __wt_row_leaf_key_info for an explanation of the magic.
+     *
+     * Only encoded keys can be upgraded to encoded key/value pairs.
      */
-    if (!(v & WT_K_FLAG)) /* Already an encoded key */
+    if ((v & WT_KEY_FLAG_BITS) != WT_K_FLAG)
         return;
 
-    key_len = WT_K_DECODE_KEY_LEN(v); /* Key length */
-    if (key_len > WT_KV_MAX_KEY_LEN)
+    if (WT_K_DECODE_KEY_CELL_OFFSET(v) > WT_KV_MAX_KEY_CELL_OFFSET) /* Key cell offset */
         return;
-    if (unpack->size > WT_KV_MAX_VALUE_LEN) /* Value length */
+    /* Not checking the prefix size, the field sizes are the same in both encodings. */
+    /* Not checking the key offset, the field sizes are the same in both encodings. */
```

Comment:
These comments look like one of them is a mistake. If you want them both, maybe reword to `Not checking the key offset or the prefix size, ...`.

Reply:
Basically, there's a set of things we do in order, and I wanted to explicitly call out that we're not doing 2 of those things. I reformatted the comment in 7f8423394 to look like this:
```
-    /* Not checking the prefix size, the field sizes are the same in both encodings. */
-    /* Not checking the key offset, the field sizes are the same in both encodings. */
+    /*
+     * Not checking the prefix size, the field sizes are the same in both encodings.
+     *
+     * Not checking the key offset, the field sizes are the same in both encodings.
+     */
```